File: examples/c/idtp_example.c
// SPDX-License-Identifier: Apache-2.0.
// Copyright (C) 2025-present idtp project and contributors.

//! IDTP usage example.

#include <idtp/idtp.h>
#include <stdint.h>
#include <string.h>
#include <stdio.h>

/// Example IDTP payload struct.
typedef struct PACKED {
    /// The value of the projection of the acceleration vector
    /// along the X axis (m/s^2).
    float acc_x;
    /// The value of the projection of the acceleration vector
    /// along the Y axis (m/s^2).
    float acc_y;
    /// The value of the projection of the acceleration vector
    /// along the Z axis (m/s^2).
    float acc_z;
    /// Angular velocity along the X axis (rad/s).
    float gyr_x;
    /// Angular velocity along the Y axis (rad/s).
    float gyr_y;
    /// Angular velocity along the Z axis (rad/s).
    float gyr_z;
} Payload;

/// Example payload size in bytes.
#define PAYLOAD_SIZE (sizeof(Payload))

/// @brief Calculate checksum for network packet.
///
/// @return Checksum for network packet.
uint16_t calculate_checksum(void) {
    // Implement this function suitable for your needs.
    return 0x1234;
}

int32_t main(void) {
    // 1) IDTP usage example - creation of raw IDTP network packet.
    // Fill custom payload with IMU sensors data.
    const Payload payload = (Payload) {
        .acc_x = 0.001f,
        .acc_y = 0.002f,
        .acc_z = 0.003f,
        .gyr_x = 0.004f,
        .gyr_y = 0.005f,
        .gyr_z = 0.006f,
    };

    const uint8_t *payload_bytes = (const uint8_t *)&payload;

    // Fill IDTP header.
    // Prefer creating IdtpHeader instance using new() method because there
    // will be no need for you to set preamble and version manually.
    IdtpHeader header = idtp_header_create();

    // Handling IDTP_MODE_SAFETY is almost the same,
    // but header.crc field should be calculated.
    header.mode = IDTP_MODE_NORMAL;
    header.device_id = 0xABCD;
    header.checksum = calculate_checksum();
    header.timestamp = 0;
    header.sequence = 0;
    header.crc = 0;
    header.payload_size = sizeof(Payload);
    header.payload_type = 0;


    // Create IDTP packet manager instance.
    IdtpFrame frame = idtp_frame_create();

    idtp_frame_set_header(&frame, &header);
    idtp_frame_set_payload(&frame, payload_bytes, PAYLOAD_SIZE);

    // Get raw network packet bytes.
    const size_t PACKET_SIZE = IDTP_PACKET_MIN_SIZE + sizeof(Payload);
    uint8_t raw_packet[PACKET_SIZE];
    memset(raw_packet, 0, PACKET_SIZE);

    idtp_frame_pack(&frame, raw_packet);
    // Handle this raw packet...

    // 2) IDTP usage example - parsing IDTP from raw network packet.

    const IdtpFrame frame2 = idtp_frame_from_bytes(raw_packet, PACKET_SIZE);
    const IdtpHeader header2 = frame2.header;

    uint8_t preamble[IDTP_PREAMBLE_SIZE + 1];
    memcpy(preamble, header2.preamble, IDTP_PREAMBLE_SIZE);
    preamble[IDTP_PREAMBLE_SIZE] = '\0';

    printf("Header preamble: %s\n", (const char *)preamble);

    uint8_t payload_bytes2[PAYLOAD_SIZE];
    memset(payload_bytes2, 0, PAYLOAD_SIZE);
    memcpy(payload_bytes2, frame.payload, PAYLOAD_SIZE);

    const Payload payload2 = *(Payload *)payload_bytes2;

    printf("Payload size: %llu\n", sizeof(payload));
    puts("Payload:");
    printf("acc_x: %lf\n", payload2.acc_x);
    printf("acc_y: %lf\n", payload2.acc_y);
    printf("acc_z: %lf\n", payload2.acc_z);
    printf("gyr_x: %lf\n", payload2.gyr_x);
    printf("gyr_y: %lf\n", payload2.gyr_y);
    printf("gyr_z: %lf\n", payload2.gyr_z);

    return 0;
}

File: examples/rust/idtp_example.rs
// SPDX-License-Identifier: Apache-2.0.
// Copyright (C) 2025-present idtp project and contributors.

//! IDTP v2.1.0 usage example.

use idtp::{
    IdtpFrame, IdtpHeader, IdtpMode,
    payload::{Imu3Acc, Imu3Gyr, Imu6},
};
use std::process;

fn main() {
    // -----------------------------------------------------------------------
    // 1) SENDER SIDE: Creating and packing an IDTP v2 frame.
    // -----------------------------------------------------------------------

    let imu_data = Imu6 {
        acc: Imu3Acc {
            acc_x: 0.001,
            acc_y: 0.002,
            acc_z: 0.003,
        },
        gyr: Imu3Gyr {
            gyr_x: 0.004,
            gyr_y: 0.005,
            gyr_z: 0.006,
        },
    };

    let mut frame = IdtpFrame::new();
    let mut header = IdtpHeader::new();

    header.mode = IdtpMode::Safety.into();
    header.device_id = 0xABCD;
    header.timestamp = 12345678;
    header.sequence = 1;

    // Important: In v2, set_header should be called before set_payload
    // or payload size must be synchronized.
    frame.set_header(&header);
    let _ = frame.set_payload(&imu_data);

    // Prepare buffer for data transmission.
    let mut buffer = [0u8; 64];

    // Packing with software-based CRC/HMAC (requires "software_impl" feature)
    // In production (MCU), use pack_with() to utilize
    // hardware CRC/HMAC accelerators.
    let packet_size = match frame.pack(&mut buffer, None) {
        Ok(size) => {
            println!("Successfully packed {} bytes", size);
            size
        }
        Err(e) => {
            eprintln!("Packing error: {:?}", e);
            process::exit(1);
        }
    };

    println!("Hex frame: {:02X?}", &buffer[..packet_size]);

    // -----------------------------------------------------------------------
    // 2) RECEIVER SIDE: Validating and parsing IDTP v2 frame.
    // -----------------------------------------------------------------------

    let incoming_data = &buffer[..packet_size];

    // Validate integrity. This checks Header CRC-8 and Frame CRC-32 without
    // creating an object.
    if let Err(e) = IdtpFrame::validate(incoming_data, None) {
        eprintln!("Invalid frame received: {:?}", e);
        return;
    }

    // Parse bytes into frame structure.
    let decoded_frame = match IdtpFrame::try_from(incoming_data) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Parse error: {:?}", e);
            return;
        }
    };

    // Extract and use data.
    let header = decoded_frame.header();

    if let Ok(payload) = decoded_frame.payload::<Imu6>() {
        println!("Received header: {:#?}", header);
        println!("Received payload: {:#?}", payload);
    }
}
